contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,447 @@
+can use AnyState to persist a history of revisions to its specification type 
+Objects. The technique is applicable, without loss of generality, to the 
+existing Kuberentes controllers that have Pod as a generated type.
+
+When a controller detects a revision to the target Object state of a 
+specification type Object it will do the following.
+
+1. The controller will [create a snapshot](#version-snapshot-creation) of the 
+current target Object state.
+1. The controller will [reconstruct the history](#history-reconstruction) of 
+revisions to the Object's target Object state.
+1. The controller will test the current target Object state for 
+[equivalence](#version-equivalence) with all other versions in the Object's 
+revision history.
+    - If the current version is semantically equivalent to its immediate 
+    predecessor no update to the Object's target state has been performed.
+    - If the current version is equivalent to a version prior to its immediate 
+    predecessor, this indicates a rollback.
+    - If the current version is not equivalent to any prior version, this 
+    indicates an update or a roll forward.
+    - Controllers should use their status objects for book keeping with respect 
+    to current and prior revisions.
+1. The controller will 
+[reconcile its generated Objects](#target-object-state-reconciliation) 
+with the new target Object state.
+1. The controller will [maintain the length of its history](#history-maintenance) 
+to be less than the configured limit.
+
+### Version Snapshot Creation
+To take a snapshot of the target Object state contained in a specification type
+Object, a controller will do the following.
+
+1. The controller will serialize all the fields of the Object's target object 
+state using JSON encoding.
+1. The controller will populate the `.Data` of an AnyState object by 
+associating the serialized fields with keys that correspond to the fields' 
+names. 
+   - e.g `PodTemplateSpecs` should be stored at the `"Template"` key.
+1. The controller will store the specification type Objects `.Generation` at the 
+`"ControllerGeneration"` key.
+1. The controller will compute the [hash](#hashing) of the AnyState's `.Data`.
+1. The controller will attach a label to the AnyState so that it is selectable 
+with a low probability of overlap.
+    - ControllerRefs will be used as the authoritative test for ownership.
+    - The specification type Object's `.Selector` should be used where 
+    applicable. 
+    - Alternatively, a Kind unique label may be set to the `.Name` of the 
+    specification type Object.
+1. The controller will add a ControllerRef indicating the specification type 
+Object as the owner of the AnyState in the AnyState's `.OwnerReferences`.
+1. The controller will use the hash from above, along with a user identifiable 
+prefix, to [generate a unique `.Name`](#unique-name-generation) for the AnyState.
```

Comment:
Each controller can decide to use whatever it needs to to prefix the state so that the name is reasonable.